RealmSwift/Projection.swift

```diff
@@ -0,0 +1,1342 @@
+import Combine
+#endif
+
+private protocol AnyProjected {
+    var projectedKeyPath: AnyKeyPath { get }
+}
+
+// MARK: Projection
+
+/// ``@Projected`` is used to declare properties on ``Projection`` protocols which should be
+/// managed by Realm.
+///
+/// Example of usage:
+/// ```swift
+/// public class Person: Object {
+///     @Persisted var firstName = ""
+///     @Persisted var lastName = ""
+///     @Persisted var address: Address? = nil
+///     @Persisted var friends = List<Person>()
+///     @Persisted var reviews = List<String>()
+/// }
+///
+/// class PersonProjection: Projection<Person> {
+///     @Projected(\Person.firstName) var firstName
+///     @Projected(\Person.lastName.localizedUppercase) var lastNameCaps
+///     @Projected(\Person.address.city) var homeCity
+///     @Projected(\Person.friends.projectTo.firstName) var firstFriendsName: ProjectedList<String>
+/// }
+///
+/// let people: Results<PersonProjection> = realm.objects(PersonProjection.self)
+/// ```
+@propertyWrapper
+public struct Projected<T: ObjectBase, Value>: AnyProjected {
+
+    fileprivate var _projectedKeyPath: KeyPath<T, Value>!
+    var projectedKeyPath: AnyKeyPath {
+        _projectedKeyPath
+    }
+
+    /// :nodoc:
+    @available(*, unavailable, message: "@Persisted can only be used as a property on a Realm object")
+    public var wrappedValue: Value {
+        // The static subscript below is called instead of this when the property
+        // wrapper is used on an ObjectBase subclass, which is the only thing we support.
+        get { fatalError("called wrappedValue getter") }
+        // swiftlint:disable:next unused_setter_value
+        set { fatalError("called wrappedValue setter") }
+    }
+
+    /// :nodoc:
+    public static subscript<EnclosingSelf: Projection<T>>(
+        _enclosingInstance observed: EnclosingSelf,
+        wrapped wrappedKeyPath: ReferenceWritableKeyPath<EnclosingSelf, Value>,
+        storage storageKeyPath: ReferenceWritableKeyPath<EnclosingSelf, Self>
+        ) -> Value {
+        get {
+            let storage = observed[keyPath: storageKeyPath]
+            return observed.rootObject[keyPath: storage._projectedKeyPath]
+        }
+        set {
+            precondition(observed[keyPath: storageKeyPath].projectedKeyPath is WritableKeyPath<T, Value>,
+                         "KeyPath is not writable")
+            observed.rootObject[keyPath: observed[keyPath: storageKeyPath].projectedKeyPath as! WritableKeyPath<T, Value>] = newValue
+        }
+    }
+
+    /// Declares a property which is lazily initialized to the type's default value.
+    public init(_ projectedKeyPath: KeyPath<T, Value>) {
+        self._projectedKeyPath = projectedKeyPath
+    }
+}
+
+// MARK: Projection Schema
+private struct ProjectedMetadata {
+    let projectedKeyPath: AnyKeyPath
+    let originPropertyKeyPathString: String
+    let label: String
+}
+
+private var schema = [ObjectIdentifier: [ProjectedMetadata]]()
```

Comment:
The code touching this needs locking. Currently initializing projections on multiple threads at once will explode.